benchmark/utils/stencil_matrix.hpp

```diff
@@ -0,0 +1,501 @@
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GINKGO_BENCHMARK_UTILS_STENCIL_MATRIX_HPP
+#define GINKGO_BENCHMARK_UTILS_STENCIL_MATRIX_HPP
+
+
+#include <ginkgo/core/base/matrix_data.hpp>
+#if GINKGO_BUILD_MPI
+#include <ginkgo/core/base/mpi.hpp>
+#endif
+
+
+template <typename T>
+double closest_nth_root(T v, int n)
+{
+    auto root = std::pow(v, 1. / static_cast<double>(n));
+    auto root_floor = std::floor(root);
+    auto root_ceil = std::ceil(root);
+    if (root - root_floor > root_ceil - root) {
+        return root_ceil;
+    } else {
+        return root_floor;
+    }
+}
+
+
+/**
+ * Generates matrix data for a 2D stencil matrix. If restricted is set to true,
+ * creates a 5-pt stencil, if it is false creates a 9-pt stencil.
+ *
+ * If
+ * strong_scaling is set to true, creates the same problem size independent of
+ * the number of ranks, if it false the problem size grows with the number of
+ * ranks.
+ */
+template <typename ValueType, typename IndexType>
+gko::matrix_data<ValueType, IndexType> generate_2d_stencil_box(
+    std::array<int, 2> dims, std::array<int, 2> position,
+    const gko::size_type target_local_size, bool restricted)
+{
+    auto num_boxes =
+        std::accumulate(dims.begin(), dims.end(), 1, std::multiplies<>{});
+
+    const auto dp =
+        static_cast<IndexType>(closest_nth_root(target_local_size, 2));
+    const auto local_size = static_cast<gko::size_type>(dp * dp);
+    const auto global_size = local_size * num_boxes;
+    auto A_data = gko::matrix_data<ValueType, IndexType>(
+        gko::dim<2>{static_cast<gko::size_type>(global_size),
+                    static_cast<gko::size_type>(global_size)});
+
+    auto global_offset = [&](const int px, const int py) {
+        return static_cast<int>(local_size) * px +
+               static_cast<int>(local_size) * dims[0] * py;
+    };
+
+    auto target_coords = [&](const IndexType i, const int coord) {
+        return 0 <= i && i <= dp - 1 ? coord : i < 0 ? coord - 1 : coord + 1;
```

Comment:
I would suggest adding parentheses for clarity
```suggestion
        return 0 <= i && (i <= dp - 1 ? coord : (i < 0 ? coord - 1 : coord + 1));
```